src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java

```diff
@@ -108,78 +102,93 @@
+                    entry(java.util.SplittableRandom.class),
+                    entry(jdk.internal.random.L32X64MixRandom.class),
+                    entry(jdk.internal.random.L64X128MixRandom.class),
+                    entry(jdk.internal.random.L64X128StarStarRandom.class),
+                    entry(jdk.internal.random.L64X256MixRandom.class),
+                    entry(jdk.internal.random.L64X1024MixRandom.class),
+                    entry(jdk.internal.random.L128X128MixRandom.class),
+                    entry(jdk.internal.random.L128X256MixRandom.class),
+                    entry(jdk.internal.random.L128X1024MixRandom.class),
+                    entry(jdk.internal.random.Xoroshiro128PlusPlus.class),
+                    entry(jdk.internal.random.Xoshiro256PlusPlus.class)
+            );
+        }
+
+        private static SimpleImmutableEntry<String, Class<? extends RandomGenerator>>
+        entry(Class<? extends RandomGenerator> rgClass) {
+            return new SimpleImmutableEntry<>(rgClass.getSimpleName(), rgClass);
         }
     }
 
     /**
      * Private constructor.
      *
-     * @param provider  Provider class to wrap.
+     * @param rgClass  Random generator class.
      */
-    private RandomGeneratorFactory(Provider<? extends RandomGenerator> provider) {
-        this.provider = provider;
+    private RandomGeneratorFactory(Class<? extends RandomGenerator> rgClass) {
+        this.rgClass = rgClass;
     }
 
     /**
      * Returns the factory map, lazily constructing map on first call.
      *
-     * @return Map of RandomGeneratorFactory classes.
+     * @return Map of random generator classes.
      */
-    private static Map<String, Provider<? extends RandomGenerator>> getFactoryMap() {
+    private static Map<String, Class<? extends RandomGenerator>> getFactoryMap() {
         return FactoryMapHolder.FACTORY_MAP;
     }
 
     /**
-     * Return the annotation for the specified provider.
+     * Return the annotation for the specified random generator.
      *
-     * @return RandomGeneratorProperties annotation for the specified provider.
+     * @return RandomGeneratorProperties annotation for the specified random generator.
      */
      private RandomGeneratorProperties getProperties() {
-        if (properties == null) {
-            synchronized (provider) {
-                if (properties == null) {
-                    properties = provider.type().getDeclaredAnnotation(RandomGeneratorProperties.class);
-                    Objects.requireNonNull(properties, provider.type() + " missing annotation");
+        if (properties == null) {  // volatile load
+            synchronized (rgClass) {
+                if (properties == null) {  // double-checking idiom
+                    RandomGeneratorProperties props = rgClass.getDeclaredAnnotation(RandomGeneratorProperties.class);
+                    Objects.requireNonNull(props, rgClass + " missing annotation");
```

Comment:
Hello Raffaello, with the `RandomGenerator` implementations now all residing within the java.base module, I think an additional advantage of that is that, it will now allow us to remove this internal `RandomGeneratorProperties` annotation and thus this reflection code.

I think one way to do that would be something like this within this `RandomGeneratorFactory` class itself: 

```java
private record RandomGenEntry(Class<?> randomGenClass, int i, int j,
                              int k, int equiDistribution, boolean stochastic,
                              boolean hardware) {

}

private static final Map<String, RandomGenEntry> FACTORY_MAP = ... // construct the map

```
where the `FACTORY_MAP` will be keyed to the alogrithm and the value will be a record which holds these additional details about the `RandomGenerator`. 
This current PR is about getting rid of ServiceLoader usage. So if you want to remove the usage of this annotation and reflection is a separate PR that's fine with me. Furthermore, although I don't see the necessity of an annotation for what we are doing here, if you think that the removal of the annotation and reflection isn't worth doing, that is OK too.

Reply:
Thinking a bit more, I think we can even get rid of the reflection used in `create()` methods implementation, if we wanted to, by doing something like this:

```java
private record RandomGenEntry(Class<? extends RandomGenerator> randomGenClass, int i, int j,
                                  int k, int equiDistribution, boolean stochastic,
                                  boolean hardware) {

        RandomGenerator create() {
            String algo = randomGenClass.getSimpleName();
            return switch (algo) {
                case "Random" -> new Random();
                case "L128X1024MixRandom" -> new L128X1024MixRandom();
                case "Xoshiro256PlusPlus" -> new Xoshiro256PlusPlus();
                // ... so on for the rest
                default -> throw new InternalError("should not happen");
            };
        }

        RandomGenerator create(long seed) {
            String algo = randomGenClass.getSimpleName();
            return switch (algo) {
                case "Random", "SplittableRandom", "SecureRandom" -> {
                    throw new UnsupportedOperationException("cannot construct with a long seed");
                }
                case "L128X1024MixRandom" -> new L128X1024MixRandom(seed);
                case "Xoshiro256PlusPlus" -> new Xoshiro256PlusPlus(seed);
                // ... so on for the rest
                default -> throw new InternalError("should not happen");
            };
        }

        RandomGenerator create(byte[] seed) {
            String algo = randomGenClass.getSimpleName();
            return switch (algo) {
                case "Random", "SplittableRandom", "SecureRandom" -> {
                    throw new UnsupportedOperationException("cannot construct with a byte[] seed");
                }
                case "L128X1024MixRandom" -> new L128X1024MixRandom(seed);
                case "Xoshiro256PlusPlus" -> new Xoshiro256PlusPlus(seed);
                // ... so on for the rest
                default -> throw new InternalError("should not happen");
            };
        }
    }

    
    private static final Map<String, RandomGenEntry> FACTORY_MAP = ... // construct the map

```

Reply:
@jaikiran I agree that we can simplify even more. I just wanted to change as little as possible in this PR to facilitate reviews.
Shall I push your proposed changes in this PR or is a followup PR preferable?

Reply:
A followup PR is fine. I think once this PR which addresses the API aspects (like removal of ServiceLoader and then updates to the create() method javadoc) is integrated, then the subsequent PR can just be all internal implementation changes like the proposed removal of reflection.

Reply:
Fine with me.

Reply:
Not sure if that's an issue - but if you wanted/needed to delay the loading of those random generator classes that will not be used until something actually wants to use them, you could consider using a `Supplier<Class<? extends RandomGenerator>>` or a `Supplier<RandomGenEntry>` for the `FACTORY_MAP` values?

Reply:
@dfuch You mean not loading the whole batch but only individual classes, as need arises?

Reply:
yes - I do not know if loading all the classes in one batch could be an issue at startup (I'd expect Random to be loaded early) - but if it proves to be, then using a Supplier to load them on demand might do the trick. If creating Random or SecureRandom does not trigger this code - and if those classes are not loaded at startup - then maybe that's a non issue and you can just ignore my comment.

Reply:
Loading of `Random`, `SecureRandom`, `SplittableRandom`, or `ThreadLocalRandom` does not trigger the loading of `RandomGeneratorFactory`.